deep_qa/data/fields/field.py

```diff
@@ -1,43 +1,62 @@
-"""
-A ``Field`` is some piece of a data instance that ends up as an array in a model (either as an
-input or an output).  Data instances are just collections of fields.
-
-
-Fields go through up to two steps of processing: (1) tokenized fields are converted into token ids,
-(2) fields containing token ids (or any other numeric data) are padded (if necessary) and converted
-into data arrays.  We have two ``Field`` objects corresponding to fields in each stage of this
-processing pipeline: ``UnindexedFields`` are fields containing raw strings (already tokenized) that
-need to be converted into token IDs.  We use these fields to compute a vocabulary, then convert
-them to ``IndexedFields``.  ``IndexedFields`` have a method for determining padding lengths, so
-that a data generator object can intelligently batch together instances, then pad them.
-
-If you are writing a new ``Field`` class, if there is nothing in the field that needs to be
-converted from strings to integers, you should just subclass ``IndexedField``.  If you `do` need to
-convert strings to integers, you should subclass both ``IndexedField`` and ``UnindexedField``, and
-make the ``UnindexedField`` subclass public, and the other private.
-"""
 from typing import Dict, List
 
 import numpy
 
 from ..vocabulary import Vocabulary
 
-# Because of how the dependencies work between these objects, we're defining them in reverse order,
-# so that UnindexedField can reference IndexedField in its return types.
-
-class IndexedField:
+class Field:
     """
-    An ``IndexedField`` is a field that is almost ready to be put into a data array for use in a
-    model.  The only thing missing is padding.  The methods on this object allow you to get padding
-    lengths, so you can group things into batches of similar size, and then actually pad the
-    field into a fixed-size array that can be used in a batch of instances.
+    A ``Field`` is some piece of a data instance that ends up as an array in a model (either as an
+    input or an output).  Data instances are just collections of fields.
+
+
+    Fields go through up to two steps of processing: (1) tokenized fields are converted into token ids,
+    (2) fields containing token ids (or any other numeric data) are padded (if necessary) and converted
+    into data arrays.  The ``Field`` object has methods to say which state the field is in, and to get
+    it from one state to the other.  If ``Field.needs_indexing()``, we will compute a vocabulary and
+    pass it to the field to use for indexing.  Once all fields are indexed, we will determine padding
+    lengths, then intelligently batch together instances and pad them into actual arrays.
     """
+    def needs_indexing(self):
```

Comment:
I do think it's reasonable for a user to implement a new `Field`, it just won't be typical, because what we have should cover most NLP uses.

I'm not sure what you're suggesting here.  You just have implementations for `count_vocab_items()` and `index()` that pass in `IndexField`?  Yeah, I could probably be convinced to do it that way.

Here are three alternatives:

1. As shown in this PR, we have a `needs_indexing` method, and `Fields` that do not need any indexing don't have to implement indexing-specific methods at all.

2. As you suggest, we remove `needs_indexing`, and just require all `Fields` to implement indexing-specific methods with a `pass`, or similar.  Because there are only two of them, this isn't that big a deal.

3. Like `SequenceField`, we introduce a new abstract `Field` type that adds the indexing-specific methods (e.g., `StringField`), and leave them out of the `Field` base class.  Then things like `IndexField` don't need to implement the indexing-specific methods, just the `Fields` that have strings, and we use an `isinstance(field, StringField)` check instead of `needs_indexing()` when iterating over `Fields` in `Instance`.  The trouble with this is it makes something like `ListField` more complicated - should it be a `StringField`?  That depends on what's inside it, really.  Also, we'd need to use multiple inheritance in order to have `TextField` based on both `SequenceField` and `StringField` (not to mention the potential for naming confusion, with too many abstract base types).

I don't have a really strong opinion between (1) and (2), though (3) seems like not a very good idea.  (1) makes the `Field` implementations slightly cleaner, with less boilerplate for stuff unrelated to a non-string field.  (2) makes the calling code cleaner, without having to check if a `Field` needs indexing.  We could even do (4) remove `needs_indexing`, and change the default `RuntimeErrors` raised in `Field` to `pass`, so that non-string fields don't need to implement anything, and the calling code is still clean...  Maybe that's the right thing to do.

What do you think?